src/main/java/org/opensearch/security/queries/QueryBuilderTraverser.java

```diff
@@ -0,0 +1,273 @@
+import org.opensearch.common.xcontent.json.JsonXContent;
+import org.opensearch.index.query.AbstractQueryBuilder;
+import org.opensearch.index.query.BoolQueryBuilder;
+import org.opensearch.index.query.BoostingQueryBuilder;
+import org.opensearch.index.query.ConstantScoreQueryBuilder;
+import org.opensearch.index.query.DisMaxQueryBuilder;
+import org.opensearch.index.query.FieldMaskingSpanQueryBuilder;
+import org.opensearch.index.query.NestedQueryBuilder;
+import org.opensearch.index.query.QueryBuilder;
+import org.opensearch.index.query.SpanContainingQueryBuilder;
+import org.opensearch.index.query.SpanFirstQueryBuilder;
+import org.opensearch.index.query.SpanMultiTermQueryBuilder;
+import org.opensearch.index.query.SpanNearQueryBuilder;
+import org.opensearch.index.query.SpanNotQueryBuilder;
+import org.opensearch.index.query.SpanOrQueryBuilder;
+import org.opensearch.index.query.SpanWithinQueryBuilder;
+import org.opensearch.index.query.functionscore.FunctionScoreQueryBuilder;
+
+public abstract class QueryBuilderTraverser {
+
+    public static boolean exists(QueryBuilder queryBuilder, Predicate<QueryBuilder> predicate) {
+        Exists traverser = new Exists(predicate);
+
+        return traverser.check(queryBuilder);
+    }
+
+    public static QueryBuilder find(QueryBuilder queryBuilder, Predicate<QueryBuilder> predicate) {
+        Exists traverser = new Exists(predicate);
+
+        if (traverser.check(queryBuilder)) {
+            return traverser.matched;
+        } else {
+            return null;
+        }
+    }
+
+    public static Set<QueryBuilder> findAll(QueryBuilder queryBuilder, Predicate<QueryBuilder> predicate) {
+        Complete traverser = new Complete(predicate);
+
+        traverser.check(queryBuilder);
+
+        return traverser.matched;
+    }
+
+    public boolean check(String query, NamedXContentRegistry namedXContentRegistry) throws IOException {
+        XContentParser parser = JsonXContent.jsonXContent.createParser(namedXContentRegistry, DeprecationHandler.THROW_UNSUPPORTED_OPERATION, query);
+        QueryBuilder queryBuilder = AbstractQueryBuilder.parseInnerQueryBuilder(parser);
+
+        return check(queryBuilder);
+    }
+
+    public abstract boolean check(QueryBuilder queryBuilder);
+
+    public abstract boolean check(List<? extends QueryBuilder> queryBuilders);
+
+    public static class Exists extends QueryBuilderTraverser {
+        private final Predicate<QueryBuilder> predicate;
+        private QueryBuilder matched;
+
+        public Exists(Predicate<QueryBuilder> predicate) {
+            this.predicate = predicate;
+        }
+
+        public boolean check(QueryBuilder queryBuilder) {
+            if (queryBuilder == null) {
+                return false;
+            }
+
+            if (predicate.test(queryBuilder)) {
+                this.matched = queryBuilder;
+                return true;
+            }
+
+            if (queryBuilder instanceof BoolQueryBuilder) {
```

Comment:
since this if/else battery is being repeated for other `QueryBuilderTraverser` implementations it might make sense to refactor it so that it only needs to be implemented once? on a superficial glance it seems to be the same for all of them?

Reply:
Not quite: Depending on the type we need to call different methods to extract the actual underlying QueryBuilders, for example:

```
constantScoreQueryBuilder.innerQuery()
disMaxQueryBuilder.innerQueries()
functionScoreQueryBuilder.query()
spanContainingQueryBuilder.bigQuery()
...
```

There is, unfortunately, no common interface etc. we could use.